src/services/textChanges.ts

```diff
@@ -883,9 +883,22 @@ namespace ts.textChanges {
                     Debug.assert(normalized[i].range.end <= normalized[i + 1].range.pos, "Changes overlap", () =>
                         `${JSON.stringify(normalized[i].range)} and ${JSON.stringify(normalized[i + 1].range)}`);
                 }
-                const textChanges = normalized.map(c =>
+                const naiveTextChanges = normalized.map(c =>
                     createTextChange(createTextSpanFromRange(c.range), computeNewText(c, sourceFile, newLineCharacter, formatContext, validate)));
-                return { fileName: sourceFile.fileName, textChanges };
+
+                // Only return changes that will actually have an impact on the file.
+                const textChanges = naiveTextChanges.filter(({ span, newText }) => {
+                    // The new text occupies a different amount of space from the old text.
+                    // It has to impact the file, so we'll keep it.
+                    if (span.length !== newText.length) {
+                        return true;
+                    }
+
+                    // The text differs, so keep it.
+                    return sourceFile.text.indexOf(newText, span.start) !== span.start;

```

Comment:
I don't know all the JS string manipulation idioms, but it seems crazy to me to search (a suffix of) the whole file when you only care about one start position.

Reply:
If it's worthwhile to avoid allocating the substring, I'd consider writing the character-by-character loop, since this seems like hot code(?).